### PR TITLE
Switch SignInClient code to task

### PR DESCRIPTION
**Pull request by mykhailo-kuchma**
Change sign in functionality from async ExecuteOrCancelled
to OlpClient CallApi. Add task to task sceduler.
Update tests.

Relates-To: OLPEDGE-2012

Signed-off-by: Mykhailo Kuchma <ext-mykhailo.kuchma@here.com>

### CODE REVIEW

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 1]
```diff
@@ -450,206 +451,124 @@ AuthenticationClient::Impl::~Impl() { pending_requests_->CancelAllAndWait(); }
...
+    client::OlpClientSettings settings;
+    settings.network_request_handler = settings_.network_request_handler;
+    settings.proxy_settings = settings_.network_proxy_settings;
+    settings.retry_settings.backdown_strategy =
+        olp::client::ExponentialBackdownStrategy();
 
-  client::CancellationContext context;
+    client::OlpClient client;
+    client.SetBaseUrl(settings_.token_endpoint_url);
+    client.SetSettings(std::move(settings));
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp:469]: Maybe move to an anonymous function?

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 2]
```diff
@@ -450,206 +451,124 @@ AuthenticationClient::Impl::~Impl() { pending_requests_->CancelAllAndWait(); }
...
-      SignInResult response(resp_impl);
-
-      if (response_status == http::HttpStatusCode::OK) {
-        // Cache the response
-        cache->locked([credentials, &response](
-                          utils::LruCache<std::string, SignInResult>& c) {
-          return c.InsertOrAssign(credentials.GetKey(), response);
-        });
+      auto cached_response = client_token_cache_->locked(
+          [credentials](utils::LruCache<std::string, SignInResult>& c)
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp:501]:
```suggestion
          [&](utils::LruCache<std::string, SignInResult>& c)
```
Sync call, we do not need to copy it.

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 3]
```diff
@@ -450,206 +451,124 @@ AuthenticationClient::Impl::~Impl() { pending_requests_->CancelAllAndWait(); }
...
-              network->Cancel(request_id);
-            }
+      return *cached_response;
+    }
+
+    auto document = std::make_shared<rapidjson::Document>();
+    rapidjson::IStreamWrapper stream(auth_response.response);
+    document->ParseStream(stream);
+
+    std::shared_ptr<SignInResultImpl> resp_impl =
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp:519]:
```suggestion
    auto resp_impl =
```

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 4]
```diff
@@ -450,206 +451,124 @@ AuthenticationClient::Impl::~Impl() { pending_requests_->CancelAllAndWait(); }
...
+        std::make_shared<SignInResultImpl>(status, "", document);
+    SignInResult response(resp_impl);
+
+    if (status == http::HttpStatusCode::OK) {
+      // Cache the response
+      client_token_cache_->locked(
+          [credentials,
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp:526]:
```suggestion
          [&
```
No need to copy on a sync call.

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp (hunk 1 above)
Comment (LiubovDidkivska) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp:470]: What do you mean about move to an anonymous function? All this code is already in sign_in_task function.

Comment (andescu) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp:470]: I mean creating an instance of OlpClient helper function.

**Pull request merged**